File: backend/app/schemas/media.py
import uuid
from datetime import datetime
from typing import Optional

from pydantic import (
    ConfigDict,
    Field,
    field_serializer,
    field_validator,
    model_validator,
)
from sqlmodel import SQLModel

from app.enums import MediaType


class PreviewPublic(SQLModel):
    """Schema for Preview public response."""
    preview_id: int
    media_id: int
    type: str
    url: str

    model_config = ConfigDict(from_attributes=True)


class AudioSettingPublic(SQLModel):
    """Schema for AudioSetting public response."""
    recording_gain_db: Optional[int] = None
    sampling_rate_hz: int
    bit_depth: Optional[int] = None
    channel_num: Optional[int] = None
    duration_s: float

    model_config = ConfigDict(from_attributes=True)


class PhotoSettingPublic(SQLModel):
    """Schema for photo technical metadata."""
    exposure_ms: Optional[float] = None
    aperture: Optional[float] = None
    iso: Optional[int] = None

    model_config = ConfigDict(from_attributes=True)


class ChunkUploadRequest(SQLModel):
    """Request for chunk upload."""
    filename: str = Field(..., description="Original filename")
    chunk_index: int = Field(..., ge=0, description="Chunk index (0-based)")
    total_chunks: int = Field(..., ge=1, description="Total number of chunks")


class ChunkUploadResponse(SQLModel):
    """Response for chunk upload."""
    filename: str
    uploaded_chunks: int
    total_chunks: int
    is_complete: bool
    file_upload_id: Optional[int] = Field(None, description="Created FileUpload ID, returned when is_complete=True")


class MediaCreate(SQLModel):
    """Request body for creating media records from uploaded files (batch)."""
    collection_id: int = Field(..., description="Collection ID")
    file_upload_ids: list[int] = Field(..., min_length=1, description="List of FileUpload IDs to process")
    filename_prefix: Optional[str] = Field(
        default=None,
        max_length=64,
        description="Optional prefix to prepend to stored filename",
    )
    date_time: Optional[str] = Field(None, description="Recording datetime in format YYYY-MM-DD HH:mm:ss")
    date_from_filename: bool = Field(False, description="Parse date from filename")
    site_id: Optional[int] = Field(None, description="Site ID")
    sensor_id: Optional[int] = Field(None, description="Sensor ID (recorder+mic combo)")
    license_id: Optional[int] = Field(None, description="License ID")
    medium: Optional[str] = Field(None, description="Medium: Air or Water")
    media_type: Optional[MediaType] = Field(None, description="Media type: audio, photo, video")
    recording_gain_db: Optional[int] = Field(None, description="Recording gain in dB")
    duty_cycle_recording: Optional[int] = Field(None, description="Duty cycle recording duration (seconds)")
    duty_cycle_period: Optional[int] = Field(None, description="Duty cycle period (seconds)")
    note: Optional[str] = Field(None, max_length=250, description="Optional note")
    doi: Optional[str] = Field(None, max_length=255, description="DOI")

    @field_validator("date_time")
    @classmethod
    def validate_date_time_format(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return None
        try:
            datetime.strptime(v, "%Y-%m-%d %H:%M:%S")
        except ValueError:
            raise ValueError("date_time must be in format YYYY-MM-DD HH:mm:ss")
        return v

    @field_validator("filename_prefix")
    @classmethod
    def validate_filename_prefix(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return None
        if ".." in v or "/" in v or "\\" in v:
            raise ValueError("filename_prefix contains invalid path characters")
        return v

    @model_validator(mode="after")
    def require_date_source(self) -> "MediaCreate":
        if self.media_type != MediaType.PHOTO and not self.date_from_filename and not self.date_time:
            raise ValueError("Either date_time or date_from_filename=true must be provided")
        if self.media_type == MediaType.PHOTO:
            audio_only_fields = {
                "recording_gain_db",
                "duty_cycle_recording",
                "duty_cycle_period",
            }
            supplied = sorted(audio_only_fields & self.model_fields_set)
            if supplied:
                raise ValueError(
                    "Photo media must not include audio-only fields: " + ", ".join(supplied)
                )
        return self


class MediaCreateFailedItem(SQLModel):
    """A failed item in batch media creation."""
    file_upload_id: int
    reason: str


class MediaCreateResponse(SQLModel):
    """Response for submitted media batch processing."""
    queue_id: int | None = Field(default=None, description="Background queue ID")
    queued: list[int] = Field(default_factory=list, description="Successfully queued file_upload_ids")
    failed: list[MediaCreateFailedItem] = Field(default_factory=list, description="Failed items with reasons")


class MediaBase(SQLModel):
    """Base schema for Media with common fields."""
    filename: Optional[str] = None
    name: Optional[str] = None
    medium: Optional[str] = None
    note: Optional[str] = None
    doi: Optional[str] = None
    date_time: Optional[datetime] = None


class MediaUpdate(SQLModel):
    """Schema for updating base Media fields; technical settings are managed elsewhere."""
    name: Optional[str] = None
    medium: Optional[str] = None
    note: Optional[str] = None
    doi: Optional[str] = None
    date_time: Optional[str] = Field(None, description="Recording datetime in format YYYY-MM-DD HH:mm:ss")
    site_id: Optional[int] = None
    sensor_id: Optional[int] = None
    license_id: Optional[int] = None
    recording_gain_db: Optional[int] = None
    sampling_rate_hz: Optional[int] = None
    bit_depth: Optional[int] = None
    channel_num: Optional[int] = None
    duration_s: Optional[float] = None
    duty_cycle_recording: Optional[int] = None
    duty_cycle_period: Optional[int] = None

    @field_validator("date_time")
    @classmethod
    def validate_date_time_format(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return None
        try:
            datetime.strptime(v, "%Y-%m-%d %H:%M:%S")
        except ValueError:
            raise ValueError("date_time must be in format YYYY-MM-DD HH:mm:ss")
        return v


class MediaListPublic(MediaBase):
    """Schema for media list responses without content-derived dimensions."""
    media_id: int
    uuid: uuid.UUID
    media_type: str
    is_metadata: bool = False
    size_b: Optional[int] = None  # Python attribute name from model (maps to size_B in DB)
    md5_hash: Optional[str] = None
    uploader_id: Optional[int] = None
    uploader_name: Optional[str] = None
    creator_id: Optional[int] = None
    creator_name: Optional[str] = None
    site_id: Optional[int] = None
    site_name: Optional[str] = None
    theme_value: Optional[str] = None
    theme_source: Optional[str] = None
    sensor_id: Optional[int] = None
    sensor_name: Optional[str] = None
    license_id: Optional[int] = None
    license_name: Optional[str] = None
    duty_cycle_recording: Optional[int] = None
    duty_cycle_period: Optional[int] = None
    creation_date: datetime

    # Related collection / project (populated in detail view)
    collection_id: Optional[int] = None
    collection_name: Optional[str] = None
    project_id: Optional[int] = None
    project_name: Optional[str] = None

    # Audio URL for player (populated in detail view)
    audio_url: Optional[str] = None
    # Original file URL for non-audio media (populated in detail view)
    media_url: Optional[str] = None
    # Previews list (populated in detail view)
    previews: list[PreviewPublic] = Field(default_factory=list)

    audio_setting: Optional[AudioSettingPublic] = None
    photo_setting: Optional[PhotoSettingPublic] = None
    labels: list[str] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)

    @field_serializer("date_time", "creation_date")
    def serialize_datetime(self, dt: Optional[datetime], _info):
        if dt is None:
            return None
            
        return dt.strftime("%Y-%m-%d %H:%M:%S")


class MediaPublic(MediaListPublic):
    """Schema for a media detail response including image dimensions."""
    image_width: Optional[int] = None
    image_height: Optional[int] = None


from app.csv_import import CsvImportResult

MetadataImportResponse = CsvImportResult


class MediaOption(SQLModel):
    """下拉选项 Schema：仅包含 media_id 与显示名称。 / Dropdown option schema: only media_id and display name."""
    media_id: int
    name: Optional[str] = None  # 优先使用 name，为空时前端可降级显示 filename
    media_type: str

    model_config = ConfigDict(from_attributes=True)


class MediaBrowseGalleryItem(SQLModel):
    """Gallery view item for media browsing."""
    media_id: int
    name: Optional[str] = None
    filename: Optional[str] = None
    media_type: str
    is_metadata: bool = False
    date_time: Optional[datetime] = None
    size_b: Optional[int] = None
    duration_s: Optional[float] = None
    sampling_rate_hz: Optional[int] = None
    bit_depth: Optional[int] = None
    channel_num: Optional[int] = None
    duty_cycle_period: Optional[int] = None
    duty_cycle_recording: Optional[int] = None
    label: Optional[str] = None
    preview_url: Optional[str] = None
    sphere: Optional[str] = None
    realm_name: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)

    @field_serializer("date_time")
    def serialize_gallery_datetime(self, dt: Optional[datetime], _info):
        if dt is None:
            return None
        return dt.strftime("%Y-%m-%d %H:%M:%S")


class MediaBrowseListItem(SQLModel):
    """List view item for media browsing."""
    media_id: int
    name: Optional[str] = None
    filename: Optional[str] = None
    media_type: str
    is_metadata: bool = False
    site_id: Optional[int] = None
    site_name: Optional[str] = None
    sensor_id: Optional[int] = None
    sensor_name: Optional[str] = None
    license_id: Optional[int] = None
    license_name: Optional[str] = None
    medium: Optional[str] = None
    date_time: Optional[datetime] = None
    duration_s: Optional[float] = None
    sampling_rate_hz: Optional[int] = None
    bit_depth: Optional[int] = None
    channel_num: Optional[int] = None
    duty_cycle_period: Optional[int] = None
    duty_cycle_recording: Optional[int] = None
    label: Optional[str] = None
    preview_url: Optional[str] = None
    size_b: Optional[int] = None
    uploader_name: Optional[str] = None
    creator_name: Optional[str] = None
    note: Optional[str] = None
    doi: Optional[str] = None
    sphere: Optional[str] = None
    topography_m: Optional[float] = None
    freshwater_depth_m: Optional[float] = None
    realm_name: Optional[str] = None
    biome_name: Optional[str] = None
    functional_type_name: Optional[str] = None
    hierarchy: list[str] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)

    @field_serializer("date_time")
    def serialize_list_datetime(self, dt: Optional[datetime], _info):
        if dt is None:
            return None
        return dt.strftime("%Y-%m-%d %H:%M:%S")


class MediaTimelineItem(SQLModel):
    """Schema for a single timeline item in media timeline view."""
    media_id: int
    media_type: str
    name: str = ""
    start_date: datetime
    end_date: datetime
    duration_s: Optional[float] = None
    site_id: Optional[int] = None
    site_key: str = "nogeo"
    site_name: str = "not geo-referenced"
    duty_cycle_period: Optional[int] = None
    duty_cycle_recording: Optional[int] = None
    is_metadata: bool = False
    creator_name: str = ""
    realm: Optional[str] = None
    item_count: int = 1

    @field_serializer("start_date", "end_date")
    @classmethod
    def serialize_datetime(cls, dt: datetime) -> str:
        return dt.strftime("%Y-%m-%d %H:%M:%S")


class MediaTimelineRange(SQLModel):
    """Schema for media timeline window range."""
    min: Optional[datetime] = None
    max: Optional[datetime] = None

    @field_serializer("min", "max")
    @classmethod
    def serialize_datetime(cls, dt: Optional[datetime]) -> Optional[str]:
        if dt is None:
            return None
        return dt.strftime("%Y-%m-%d %H:%M:%S")


class MediaTimelineResponse(SQLModel):
    """Schema for media timeline payload."""
    project_id: int
    collection_id: Optional[int] = None
    items: list[MediaTimelineItem] = Field(default_factory=list)
    time_range: MediaTimelineRange
    has_more: bool = False


class MediaCollectionLinksSyncRequest(SQLModel):
    """Request schema for syncing collection links across multiple media records."""

    media_ids: list[int] = Field(..., min_length=1, description="Media IDs to update")
    collection_ids: list[int] = Field(default_factory=list, description="Collection IDs to set on every media")

    model_config = ConfigDict(extra="forbid")


class MediaBatchFailedItem(SQLModel):
    """Failed item returned by batch media operations."""

    media_id: int
    status_code: int
    message: str


class MediaBatchOperationResponse(SQLModel):
    """Common response for batch media relation operations."""

    succeeded: list[int] = Field(default_factory=list)
    failed: list[MediaBatchFailedItem] = Field(default_factory=list)


class MediaLinkCollectionItem(SQLModel):
    """Collection item used by media-collection link dialog."""
    collection_id: int
    name: str
    selected: bool = False
    duplicate_project_ids: list[int] = Field(default_factory=list)


class MediaLinkCurrentProject(SQLModel):
    """Current project block for media-collection link dialog."""
    project_id: int
    project_name: str
    collections: list[MediaLinkCollectionItem] = Field(default_factory=list)


class MediaLinkOtherProject(SQLModel):
    """Other project block for media-collection link dialog."""
    project_id: int
    project_name: str
    collections: list[MediaLinkCollectionItem] = Field(default_factory=list)


class MediaCollectionLinkOptionsResponse(SQLModel):
    """Response schema for media-collection link dialog options."""
    current_project: MediaLinkCurrentProject
    other_projects: list[MediaLinkOtherProject] = Field(default_factory=list)
    unassigned_collections: list[MediaLinkCollectionItem] = Field(default_factory=list)
    selected_collection_ids: list[int] = Field(default_factory=list)


class MediaNavigationItem(SQLModel):
    """A single media item used in navigation response."""
    media_id: int
    name: Optional[str] = None


class MediaNavigation(SQLModel):
    """Navigation response for prev/next media within the same collection."""
    prev: Optional[MediaNavigationItem] = None
    next: Optional[MediaNavigationItem] = None

File: frontend/src/api/endpoints/media.ts
import { apiClient } from "../client"
import { getApiData } from "../utils"
import type { FilterOptionLabel, FilterOptionUser } from "../utils"

export interface MediaPublic {
    media_id?: number
    uuid?: string
    media_type: string
    is_metadata?: boolean
    name?: string | null
    filename?: string | null
    site_id?: number | null
    site_name?: string | null
    sensor_id?: number | null
    sensor_name?: string | null
    medium?: string | null
    bit_depth?: string | number
    size_b?: number | null
    license_id?: number | null
    license_name?: string | null
    doi?: string | null
    note?: string | null
    uploader_id?: number | null
    uploader_name?: string | null
    creator_id?: number | null
    creator_name?: string | null
    /** browse/list 接口返回 date_time */
    date_time?: string
    duty_cycle_recording?: number | null
    duty_cycle_period?: number | null
    label?: string | null
    labels?: string[]
    hierarchy?: string[]
    preview_url?: string | null
    spectrogram?: string | null
    sphere?: string | null
    realm_name?: string | null
    site_realm_name?: string | null
    theme_value?: string | null
    theme_source?: string | null
    topography_m?: number | null
    freshwater_depth_m?: number | null
    audio_setting?: MediaAudioSetting | null
    photo_setting?: MediaPhotoSetting | null
    media_url?: string | null
    image_width?: number | null
    image_height?: number | null
    [key: string]: any
}

export interface MediaListItem extends MediaPublic {
    media_id: number
    labels: string[]
}

export interface GetMediaParams {
    page?: number
    page_size?: number
    order_by?: string
    order_dir?: "asc" | "desc"
    project_id?: number
    collection_id?: number
    [key: string]: any
}

export interface PagedMediaResponse<T = MediaPublic> {
    code: number
    message: string
    data: T[]
    page_info: {
        total: number
        page: number
        page_size: number
        total_pages: number
    }
}


interface CreateMediaPayloadBase {
    collection_id: number
    file_upload_ids: number[]
    filename_prefix?: string
    date_time?: string
    date_from_filename?: boolean
    site_id?: number
    sensor_id?: number
    creator_id?: number | null
    license_id?: number
    medium?: string
    note?: string
    doi?: string
}

export interface CreateAudioMediaPayload extends CreateMediaPayloadBase {
    media_type: "audio"
    recording_gain_db?: number
    duty_cycle_recording?: number
    duty_cycle_period?: number
}

export interface CreatePhotoMediaPayload extends CreateMediaPayloadBase {
    media_type: "photo"
    recording_gain_db?: never
    duty_cycle_recording?: never
    duty_cycle_period?: never
}

/** Creation fields are deliberately discriminated so a photo cannot carry audio settings. */
export type CreateMediaPayload = CreateAudioMediaPayload | CreatePhotoMediaPayload

export interface UpdateMediaPayload {
    name?: string | null
    date_time?: string | null
    site_id?: number | null
    sensor_id?: number | null
    medium?: string | null
    license_id?: number | null
    creator_id?: number | null
    doi?: string | null
    note?: string | null
    recording_gain_db?: number | null
    sampling_rate_hz?: number | null
    bit_depth?: number | null
    channel_num?: number | null
    duration_s?: number | null
    duty_cycle_recording?: number | null
    duty_cycle_period?: number | null
}

export interface CreateMediaParams {
    project_id?: number
    [key: string]: string | number | boolean | undefined
}

export interface MediaCreateFailedItem {
    file_upload_id: number
    reason: string
}

/** Response body of `POST /v1/media`. */
export interface MediaCreateResponse {
    queue_id?: number | null
    queued: number[]
    failed: MediaCreateFailedItem[]
}

export interface BrowseMediaParams {
    project_id: number
    view_type: "gallery" | "list"
    page?: number
    page_size?: number
    collection_id?: number | null
    name?: string | null
    media_type?: "audio" | "photo" | null
    order_by?: string | null
    order_dir?: "asc" | "desc" | null
}

/** Detail payload returned by GET /v1/media/{id}. */
export interface MediaPreviewItem {
    preview_id: number
    media_id: number
    type: string
    url: string
}

export interface MediaAudioSetting {
    recording_gain_db?: number
    sampling_rate_hz?: number
    bit_depth?: number
    channel_num?: number
    duration_s?: number
}

export interface MediaPhotoSetting {
    exposure_ms?: number | null
    aperture?: number | null
    iso?: number | null
}

export interface RecordingDetail extends Record<string, unknown> {
    media_id?: number
    id?: number
    is_metadata?: boolean
    filename?: string
    name?: string
    audio_url?: string
    media_url?: string
    audio_setting?: MediaAudioSetting
    photo_setting?: MediaPhotoSetting
    image_width?: number
    image_height?: number
    previews?: MediaPreviewItem[]
    collection_id?: number
    project_id?: number
    labels?: unknown[]
    label?: string
    uuid?: string
    note?: string
    creation_date?: string
    site_name?: string
    sensor_name?: string
    medium?: string
    license_name?: string
    doi?: string
    creator_id?: number
    creator_name?: string
    uploader_id?: number
    uploader_name?: string
    spectrogram?: string
    duration_s?: number
    sampling_rate_hz?: number
    theme_value?: string | null
    theme_source?: string | null
    channels?: string
    bit_depth?: string
    size_b?: number
    gain?: string
    download_url?: string
    date_time?: string
}

export const RECORDING_FFT_SIZES = [128, 256, 512, 1024, 2048, 4096] as const

export type SpectrogramQueryParams = {
    start_time?: number
    end_time?: number
    min_freq?: number
    max_freq?: number
    fft_size?: number
    window?: string
    channel?: number
    width?: number
    height?: number
    filter?: boolean
}

export type MediaAudioQueryParams = {
    start_time?: number
    end_time?: number
    min_freq?: number
    max_freq?: number
    fft_size?: number
    filter?: boolean
    /** 前端 reload token，避免浏览器 disk cache 命中错误频段 */
    reload_key?: number
}

/** GET /v1/media-options — 下拉用（media_id + 显示名） */
export interface MediaOption {
    media_id: number
    name?: string | null
    media_type?: string | null
    is_metadata?: boolean
    filename?: string | null
}

export const mediaApi = {
    /** 获取 Media 列表 */
    getMedia(params?: GetMediaParams) {
        const cleanParams = params ? Object.fromEntries(Object.entries(params).filter(([_, v]) => v !== undefined && v !== '')) : undefined
        return apiClient.get<PagedMediaResponse<MediaListItem>>("/v1/media", { params: cleanParams as any })
    },

    /** 浏览 Media 列表（支持画廊/列表视图） */
    browseMedia(params: BrowseMediaParams, ignoreUnauthorized?: boolean) {
        const cleanParams = Object.fromEntries(
            Object.entries(params).filter(([_, v]) => v !== undefined && v !== null && v !== ""),
        )
        return apiClient.get<PagedMediaResponse>("/v1/media-browse-items", {
            params: cleanParams as any,
            ignoreUnauthorized,
        })
    },

    /** 批量创建媒体记录 */
    createMedia(payload: CreateMediaPayload, params?: CreateMediaParams) {
        return apiClient.post<{ code: number; message: string; data: MediaCreateResponse }>("/v1/media", payload, { params })
    },

    /** 获取媒体记录详情 */
    getMediaDetail(id: number, projectId: number, ignoreUnauthorized?: boolean) {
        return apiClient.get<{ code: number; message: string; data: MediaPublic }>(
            `/v1/media/${id}`,
            { params: { project_id: projectId }, ignoreUnauthorized },
        )
    },

    /** 详情页：校验 code===0 并返回 data */
    async getRecordingDetail(mediaId: number, projectId: number, ignoreUnauthorized?: boolean): Promise<RecordingDetail> {
        const res = await apiClient.get<{ code: number; message: string; data: RecordingDetail }>(
            `/v1/media/${mediaId}`,
            { params: { project_id: projectId }, ignoreUnauthorized },
        )
        return getApiData(res)
    },

    getMediaContent(mediaId: number, projectId: number) {
        return apiClient.download(`/v1/media/${mediaId}/content`, { params: { project_id: projectId } })
    },

    /** GET /v1/media-options — 顶栏录音切换等 */
    async getMediaOptions(
        params: {
            project_id: number
            collection_id?: number
            name?: string
        },
        ignoreUnauthorized?: boolean,
    ): Promise<MediaOption[]> {
        const clean = Object.fromEntries(
            Object.entries(params).filter(([, v]) => v !== undefined && v !== null && v !== ""),
        ) as Record<string, string | number>
        const res = await apiClient.get<{ code: number; message: string; data: MediaOption[] }>(
            "/v1/media-options",
            { params: clean, ignoreUnauthorized },
        )
        return getApiData(res)
    },

    /** 实时声谱图 PNG（需 Bearer，用 blob + object URL 展示） */
    fetchSpectrogramBlob(mediaId: number, projectId: number, params?: SpectrogramQueryParams, ignoreUnauthorized?: boolean) {
        const clean = params
            ? (Object.fromEntries(
                  Object.entries(params).filter(([, v]) => v !== undefined && v !== null),
              ) as Record<string, string | number | boolean | undefined>)
            : undefined
        return apiClient.download(`/v1/media/${mediaId}/spectrogram`, {
            params: { ...(clean ?? {}), project_id: projectId },
            ignoreUnauthorized,
        })
    },

    /** 音频流（整段拉取为 blob，便于带 Authorization；大文件注意内存） */
    fetchAudioBlob(mediaId: number, projectId: number, params?: MediaAudioQueryParams, ignoreUnauthorized?: boolean) {
        const clean = params
            ? (Object.fromEntries(
                  Object.entries(params).filter(([, v]) => v !== undefined && v !== null),
              ) as Record<string, string | number | boolean | undefined>)
            : undefined
        return apiClient.download(`/v1/media/${mediaId}/audio`, {
            params: { ...(clean ?? {}), project_id: projectId },
            ignoreUnauthorized,
        })
    },

    /** 更新媒体记录 */
    updateMedia(id: number, projectId: number, payload: UpdateMediaPayload) {
        return apiClient.patch<{ code: number; message: string; data: MediaPublic }>(
            `/v1/media/${id}`,
            payload,
            { params: { project_id: projectId } },
        )
    },

    /** 删除媒体记录 */
    deleteMedia(id: number) {
        return apiClient.delete<{ code: number; message: string; data: any }>(`/v1/media/${id}`)
    },

    /** 获取媒体关联集合弹窗数据 */
    getCollectionLinkOptions(mediaId: number, params: { project_id: number; name?: string; other_project_name?: string }) {
        return apiClient.get<{ code: number; message: string; data: any }>(`/v1/media/${mediaId}/collection-options`, {
            params,
        })
    },

    /** 批量更新媒体的集合关联关系 */
    updateMediaCollectionLinks(mediaIds: number[], projectId: number, collectionIds: number[]) {
        return apiClient.put<{ code: number; message: string; data: any }>(
            "/v1/media-collection-links",
            { media_ids: mediaIds, collection_ids: collectionIds },
            { params: { project_id: projectId } },
        )
    },

    exportAudioCsv(params: GetMediaParams) {
        const { project_id, collection_id, order_by, order_dir } = params
        return apiClient.download("/v1/audios/exports", {
            params: { project_id, collection_id, order_by, order_dir },
        })
    },

    exportPhotoCsv(params: GetMediaParams) {
        const { project_id, collection_id, order_by, order_dir } = params
        return apiClient.download("/v1/photos/exports", {
            params: { project_id, collection_id, order_by, order_dir },
        })
    },

    /** Data > Audios：Uploader / Creator 列筛选下拉（project_id 必填） */
    getFilterOptions(params: { project_id: number; collection_id?: number }) {
        const cleanParams = Object.fromEntries(
            Object.entries(params).filter(([_, v]) => v !== undefined && v !== null),
        ) as { project_id: number; collection_id?: number }
        return apiClient.get<{
            code: number
            message: string
            data: { uploader: FilterOptionUser[]; creator: FilterOptionUser[]; labels: FilterOptionLabel[] }
        }>("/v1/media/filter-options", { params: cleanParams as any })
    },
}
